### PR TITLE
Remove testcase specific encoding

### PR DESCRIPTION
**Pull request by dstallenberg**
The test case encoding has been moved to the individual tools as it was too dependent on the specifics of the language.

### CODE REVIEW

File: src/search/Decoder.ts [hunk 1]
```diff
@@ -16,36 +16,24 @@
...
 
 /**
- * TestCaseDecoder interface.
+ * Decoder interface.
  *
  * @author Dimitri Stallenberg
  * @author Mitchell Olsthoorn
  */
-export interface TestCaseDecoder {
+export interface Decoder<E extends Encoding, D> {
```
Comment (mitchellolsthoorn) [inline src/search/Decoder.ts:27]: Add comment about what the generics mean

File: src/search/Decoder.ts [hunk 2]
```diff
@@ -16,36 +16,24 @@
...
-   * @param testCase             the testCase to decode
-   * @param addLogs              whether to add log statements in the test case
-   * @param additionalAssertions a dictionary of additional assertions to put in the test case
-   * @param targetName           the name of the target, used to create informative test cases
-   * @return                     the decoded test case
+   * @param encoding             the encoding to decode
+   * @param targetName           the name of the target, used to create informative decodings
+   * @return decoded             the decoded encoding
    */
   decodeTestCase(
```
Comment (mitchellolsthoorn) [inline src/search/Decoder.ts:35]: rename to decode

File: src/search/Encoding.ts [hunk 3]
```diff
@@ -18,38 +18,130 @@
...
+
+  get id(): string {
+    return this._id;
+  }
+
+  get assertions(): Map<string, string> {
+    return this._assertions;
+  }
+
+  set assertions(value: Map<string, string>) {
```
Comment (mitchellolsthoorn) [inline src/search/Encoding.ts:90]: Is not immutable

Comment (dstallenberg) [reply src/search/Encoding.ts:90]: Currently it is not possible to change this
For now, the Encoding does have mutable parts.
The genes of the encoding are immutable

File: src/search/Decoder.ts (hunk 2 above)
Comment (dstallenberg) [reply src/search/Decoder.ts]: Done

File: src/search/Decoder.ts (hunk 1 above)
Comment (dstallenberg) [reply src/search/Decoder.ts:30]: Done

**Review by mitchellolsthoorn — APPROVED** [review]

**Review by mitchellolsthoorn — APPROVED** [review]

**Review by mitchellolsthoorn — DISMISSED** [review]

**Review by mitchellolsthoorn — APPROVED** [review]

**Pull request merged**